ambersim/base.py

```diff
@@ -0,0 +1,100 @@
+import jax
+from jax import numpy as jp
+import numpy as np
+from typing import Tuple, Dict, Any
+
+from brax.base import Base, Motion, Transform
+from brax.envs.base import Env
+import mujoco
+from mujoco import mjx
+from flax import struct
+
+
+class MjxEnv(Env):
+    """API for driving an MJX system for training and inference in brax."""
+
+    def __init__(
+        self,
+        mj_model: mujoco.MjModel,
+        physics_steps_per_control_step: int = 1,
+    ):
+        """Initializes MjxEnv.
+
+        Args:
+          mj_model: mujoco.MjModel
+          physics_steps_per_control_step: the number of times to step the physics
+            pipeline for each environment step
+        """
+        self.model = mj_model
+        self.data = mujoco.MjData(mj_model)
+        self.sys = mjx.device_put(mj_model)
+        self._physics_steps_per_control_step = physics_steps_per_control_step
+
+    def pipeline_init(self, qpos: jax.Array, qvel: jax.Array) -> mjx.Data:
+        """Initializes the physics state."""
+        data = mjx.device_put(self.data)
+        data = data.replace(qpos=qpos, qvel=qvel, ctrl=jp.zeros(self.sys.nu))
+        data = mjx.forward(self.sys, data)
+        return data
+
+    def pipeline_step(self, data: mjx.Data, ctrl: jax.Array) -> mjx.Data:
+        """Takes a physics step using the physics pipeline."""
+
+        def f(data, _):
+            data = data.replace(ctrl=ctrl)
+            return (
+                mjx.step(self.sys, data),
+                None,
+            )
+
+        data, _ = jax.lax.scan(f, data, (), self._physics_steps_per_control_step)
+        return data
+
+    @property
+    def dt(self) -> jax.Array:
+        """The timestep used for each env step."""
+        return self.sys.opt.timestep * self._physics_steps_per_control_step
+
+    @property
+    def observation_size(self) -> int:
+        rng = jax.random.PRNGKey(0)
+        reset_state = self.unwrapped.reset(rng)
+        return reset_state.obs.shape[-1]
+
+    @property
+    def action_size(self) -> int:
+        return self.sys.nu
+
+    @property
+    def backend(self) -> str:
```

Comment:
it might be helpful to specify here what backends `mjx` accepts elsewhere so people who might use different backends know what's admissible before overwriting this method.

Reply:
I think this backend is just there as a placeholder for it to be in similar format as brax (since brax has several backends you could choose from).